sk-gen/src/main.rs

```diff
@@ -0,0 +1,700 @@
+    deployments: BTreeMap<String, Deployment>,
+}
+
+impl Node {
+    /// Creates a new state with no active [`Deployment`]s.
+    ///
+    /// This can be revised in future to, for instance, start at the end of an existing trace.
+    fn new() -> Self {
+        Self { deployments: BTreeMap::new() }
+    }
+
+    /// Attempts to create a [`Deployment`] in this state.
+    ///
+    /// Returns [`None`] if the deployment already exists.
+    fn create_deployment(&self, name: &str, candidate_deployments: &BTreeMap<String, Deployment>) -> Option<Self> {
+        let deployment = candidate_deployments.get(name)?;
+
+        let mut next_state = self.clone();
+        next_state.deployments.insert(name.to_string(), deployment.clone());
+        Some(next_state)
+    }
+
+    /// Attempts to delete a [`Deployment`] from this state.
+    ///
+    /// Returns [`None`] if the deployment does not exist.
+    fn delete_deployment(&self, name: &str) -> Option<Self> {
+        if self.deployments.contains_key(name) {
+            let mut next_state = self.clone();
+            next_state.deployments.remove(name);
+            Some(next_state)
+        } else {
+            None
+        }
+    }
+
+    /// Attempts to increment the replica count of an active [`Deployment`] in this state.
+    ///
+    /// Returns [`None`] if the deployment does not exist.
+    fn increment_replica_count(&self, name: String) -> Option<Self> {
+        let incremented_deployment = self.deployments.get(&name)?.increment()?;
+
+        let mut next_state = self.clone();
+        next_state.deployments.insert(name, incremented_deployment);
+        Some(next_state)
+    }
+
+    /// Attempts to decrement the replica count of an active [`Deployment`] in this state.
+    ///
+    /// Returns [`None`] if the deployment does not exist.
+    fn decrement_replica_count(&self, name: String) -> Option<Self> {
+        let decremented_deployment = self.deployments.get(&name)?.decrement()?;
+
+        let mut next_state = self.clone();
+        next_state.deployments.insert(name, decremented_deployment);
+        Some(next_state)
+    }
+
+    /// Attempts to perform a [`ClusterAction`] on this [`Node`] to obtain a next [`Node`].
+    ///
+    /// Returns [`None`] if the action is invalid.
+    fn perform_action(
+        &self,
+        ClusterAction { target_name: name, action_type }: ClusterAction,
+        candidate_deployments: &BTreeMap<String, Deployment>,
+    ) -> Option<Self> {
+        match action_type {
+            DeploymentAction::IncrementReplicas => self.increment_replica_count(name),
+            DeploymentAction::DecrementReplicas => self.decrement_replica_count(name),
+            DeploymentAction::CreateDeployment => self.create_deployment(&name, candidate_deployments),
+            DeploymentAction::DeleteDeployment => self.delete_deployment(&name),
+        }
+    }
+
+    /// Enumerates at least all possible `ClusterAction` instances.
```

Comment:
This makes sense but wondering if it is efficient enough - do we anticipate performance issues of computing all possible and then eliminating vs eliminating on the fly?

Reply:
Potentially! And this isolation of generation/execution makes it very easy to make requisite changes to that end!